app/services/BusinessRatesAttachmentsService.scala

```diff
@@ -72,13 +73,48 @@ class BusinessRatesAttachmentsService @Inject()(
 
   def submit(
               submissionId: String,
-              nonEmptyReferences: List[String]
+              nonEmptyReferences: List[String],
+              retryCount: Int = 0
             )(implicit request: LinkingSessionRequest[_], hc: HeaderCarrier): EitherT[Future, AttachmentException, List[Attachment]] = {
     EitherT.fromEither[Future](Either.cond(nonEmptyReferences.nonEmpty, nonEmptyReferences, MissingRequiredNumberOfFiles))
       .semiflatMap(Future.traverse(_)(r => getAttachment(r).map(r -> _)))
+      .subflatMap(attachments => {
+        Either.cond(
+          attachments.filter(
+            attachment =>
+              List(
+                UploadPending,
+                Uploading,
+                UploadAttachmentFailed,
+                UploadAttachmentComplete,
+                UploadingScanResults,
+                UploadScanResultsFailed,
+                UploadScanResultsComplete
+              ).contains(attachment._2.state)).size != nonEmptyReferences.size, attachments, AllFilesAreAlreadyUploaded(attachments.map(_._2)))
+      }
+      )
+      .subflatMap { attachments =>
+        val (uploaded, notuploaded) =
+          attachments.partition(attachment =>
+          List(
+          UploadPending,
+          Uploading,
+          UploadAttachmentFailed,
+          UploadAttachmentComplete,
+          UploadingScanResults,
+          UploadScanResultsFailed,
+          UploadScanResultsComplete
+        ).contains(attachment._2.state))
+        Either.cond(uploaded.isEmpty, attachments, SomeFilesAreAlreadyUploaded(notuploaded.map(_._1)))
+      }
       .map(result => result.filter(_._2.state == MetadataPending))
       .subflatMap(filtered => Either.cond(filtered.size == nonEmptyReferences.size, filtered.map(_._1), NotAllFilesReadyToUpload))
       .semiflatMap(references => Future.traverse(references)(patchMetadata(submissionId, _)))
+      .leftFlatMap{
```

Comment:
you can move the if-condition to the left of the arrow and then you don't need the else clause and the repeated leftT(error)

```scala
.leftFlatMap{
        case AllFilesAreAlreadyUploaded(attachments)  => EitherT.rightT(attachments)
        case error @SomeFilesAreAlreadyUploaded(references)  if (retryCount < 5) =>  submit(submissionId, references, retryCount + 1)
        case error                                    =>    EitherT.leftT(error)
      }
```